tokio-net/src/driver/sharded_slab/mod.rs

```diff
@@ -0,0 +1,307 @@
+//! A lock-free concurrent slab.
+mod page;
+mod tid;
+pub(crate) use tid::Tid;
+mod iter;
+
+use crate::sync::{
+    atomic::{AtomicUsize, Ordering},
+    CausalCell,
+};
+use page::{slot, Addr, Page};
+use std::fmt;
+
+#[cfg(target_pointer_width = "64")]
+const MAX_THREADS: usize = 4096;
+#[cfg(target_pointer_width = "32")]
+const MAX_THREADS: usize = 2048;
+
+const MAX_PAGES: usize = 16;
+const RESERVED_BITS: usize = 5;
+
+/// A sharded, lock-free slab.
+pub(crate) struct Slab<T> {
+    shards: Box<[CausalCell<Shard<T>>]>,
```

Comment:
Since the maximum number of shards is hard-coded to a fixed constant, this could conceivably be replaced by an array, removing a pointer indirection. I didn't want to mess around with `MaybeUninit` so I punted on that for now, but it might be worth looking into as a follow-up change.